central/baseimage/datastore/datastore_impl.go

```diff
@@ -0,0 +1,155 @@
+package datastore
+
+import (
+	"context"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/pkg/errors"
+	baseImageStore "github.com/stackrox/rox/central/baseimage/store/postgres"
+	"github.com/stackrox/rox/generated/storage"
+	"github.com/stackrox/rox/pkg/logging"
+	"github.com/stackrox/rox/pkg/postgres"
+	"github.com/stackrox/rox/pkg/sac"
+	"github.com/stackrox/rox/pkg/sac/resources"
+	"github.com/stackrox/rox/pkg/uuid"
+)
+
+type datastoreImpl struct {
+	storage baseImageStore.Store
+	db      postgres.DB
+}
+
+var (
+	log       = logging.LoggerForModule()
+	imagesSAC = sac.ForResource(resources.ImageAdministration)
+)
+
+const (
+	baseImagesTable = "base_images"
+	// The 'firstlayerdigest' column is indexed.
+	listByFirstLayerQuery = "SELECT id FROM " + baseImagesTable + " WHERE firstlayerdigest = $1"
+
+	// Query to find the ID by manifest digest
+	getByManifestDigestQuery = "SELECT id FROM " + baseImagesTable + " WHERE manifestdigest = $1"
+)
+
+// New creates a new DataStore instance backed by PostgreSQL.
+func New(store baseImageStore.Store, db postgres.DB) DataStore {
+	return &datastoreImpl{
+		storage: store,
+		db:      db,
+	}
+}
+
+func (ds *datastoreImpl) UpsertImage(ctx context.Context, image *storage.BaseImage, layers []*storage.BaseImageLayer) error {
+	if ok, err := imagesSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	for i, layer := range layers {
+		if layer.GetId() == "" {
+			layer.Id = uuid.NewV4().String()
+		}
+
+		layer.BaseImageId = image.GetId()
+		layer.Index = int32(i)
```

Comment:
The only field being pulled from the `storage.BaseImageLayer` object is the `id` and `layer_digest` -- and `id` is actually optional. Can we accept a list of layer digests in this datastore function?